File: sdks/php/src/Management/Management.php
<?php

declare(strict_types=1);

namespace Doow\Track\Management;

use Doow\Track\DoowError;
use Doow\Track\App;
use Doow\Track\Contract;
use Doow\Track\License;
use Doow\Track\Metric;
use Doow\Track\Expense;
use Doow\Track\PaginatedResponse;
use GuzzleHttp\Client;

class ManagementOptions
{
    public function __construct(
        public string $endpoint = 'https://api.doow.co',
        public int $timeoutMs = 30000,
        public bool $debug = false,
    ) {
        if ($env = getenv('DOOW_TRACK_ENDPOINT')) {
            $this->endpoint = $env;
        }
    }
}

class Management
{
    private const SDK_VERSION = '0.1.0';

    private string $apiKey;
    private ManagementOptions $options;
    private Client $client;

    public function __construct(string $apiKey, ?ManagementOptions $options = null)
    {
        $this->apiKey = getenv('DOOW_TRACK_API_KEY') ?: $apiKey;
        $this->options = $options ?? new ManagementOptions();

        $this->client = new Client([
            'timeout' => $this->options->timeoutMs / 1000,
        ]);
    }

    private function log(string $message): void
    {
        if ($this->options->debug) {
            fwrite(STDERR, "[doow/management] {$message}\n");
        }
    }

    private function request(string $method, string $path, ?array $body = null): array
    {
        $this->log("{$method} {$path}");

        $options = [
            'headers' => [
                'Authorization' => "Bearer {$this->apiKey}",
                'Content-Type' => 'application/json',
                'User-Agent' => 'doow-track-php/' . self::SDK_VERSION,
            ],
            'http_errors' => false,
        ];

        if ($body !== null) {
            $options['json'] = $body;
        }

        $response = $this->client->request(
            $method,
            "{$this->options->endpoint}{$path}",
            $options
        );

        $statusCode = $response->getStatusCode();
        $contents = $response->getBody()->getContents();

        if ($statusCode >= 400) {
            $data = json_decode($contents, true) ?? [];
            throw new DoowError(
                status: $statusCode,
                message: $data['message'] ?? 'Unknown error',
                errorClass: $data['errorClass'] ?? null,
                details: $data,
            );
        }

        if ($statusCode === 204 || empty($contents)) {
            return [];
        }

        return json_decode($contents, true) ?? [];
    }

    public function apps(): AppsResource
    {
        return new AppsResource($this);
    }

    public function contracts(): ContractsResource
    {
        return new ContractsResource($this);
    }

    public function licenses(): LicensesResource
    {
        return new LicensesResource($this);
    }

    public function metrics(): MetricsResource
    {
        return new MetricsResource($this);
    }

    public function expenses(): ExpensesResource
    {
        return new ExpensesResource($this);
    }

    public function get(string $path): array
    {
        return $this->request('GET', $path);
    }

    public function post(string $path, array $body): array
    {
        return $this->request('POST', $path, $body);
    }

    public function patch(string $path, array $body): array
    {
        return $this->request('PATCH', $path, $body);
    }

    public function delete(string $path): void
    {
        $this->request('DELETE', $path);
    }
}

class AppsResource
{
    public function __construct(private Management $client) {}

    public function list(int $limit = 25, ?string $cursor = null): PaginatedResponse
    {
        $query = "?limit={$limit}";
        if ($cursor) {
            $query .= "&cursor={$cursor}";
        }
        $data = $this->client->get("/sdk/apps{$query}");
        return new PaginatedResponse(
            data: array_map(fn($a) => App::fromArray($a), $data['data'] ?? []),
            nextCursor: $data['next_cursor'] ?? null,
            hasMore: $data['has_more'] ?? false,
        );
    }

    public function get(string $appId): App
    {
        $data = $this->client->get("/sdk/apps/{$appId}");
        return App::fromArray($data);
    }

    public function create(array $input): App
    {
        $data = $this->client->post('/sdk/apps', $input);
        return App::fromArray($data);
    }

    public function update(string $appId, array $input): App
    {
        $data = $this->client->patch("/sdk/apps/{$appId}", $input);
        return App::fromArray($data);
    }

    public function delete(string $appId): void
    {
        $this->client->delete("/sdk/apps/{$appId}");
    }
}

class ContractsResource
{
    public function __construct(private Management $client) {}

    public function list(string $appId, int $limit = 25, ?string $cursor = null): PaginatedResponse
    {
        $query = "?limit={$limit}";
        if ($cursor) {
            $query .= "&cursor={$cursor}";
        }
        $data = $this->client->get("/sdk/apps/{$appId}/contracts{$query}");
        return new PaginatedResponse(
            data: array_map(fn($c) => Contract::fromArray($c), $data['data'] ?? []),
            nextCursor: $data['next_cursor'] ?? null,
            hasMore: $data['has_more'] ?? false,
        );
    }

    public function get(string $contractId): Contract
    {
        $data = $this->client->get("/sdk/contracts/{$contractId}");
        return Contract::fromArray($data);
    }

    public function create(string $appId, array $input): Contract
    {
        $data = $this->client->post("/sdk/apps/{$appId}/contracts", $input);
        return Contract::fromArray($data);
    }

    public function update(string $contractId, array $input): Contract
    {
        $data = $this->client->patch("/sdk/contracts/{$contractId}", $input);
        return Contract::fromArray($data);
    }

    public function delete(string $contractId): void
    {
        $this->client->delete("/sdk/contracts/{$contractId}");
    }
}

class LicensesResource
{
    public function __construct(private Management $client) {}

    public function list(string $contractId, int $limit = 25, ?string $cursor = null): PaginatedResponse
    {
        $query = "?limit={$limit}";
        if ($cursor) {
            $query .= "&cursor={$cursor}";
        }
        $data = $this->client->get("/sdk/contracts/{$contractId}/licenses{$query}");
        return new PaginatedResponse(
            data: array_map(fn($l) => License::fromArray($l), $data['data'] ?? []),
            nextCursor: $data['next_cursor'] ?? null,
            hasMore: $data['has_more'] ?? false,
        );
    }

    public function get(string $licenseId): License
    {
        $data = $this->client->get("/sdk/licenses/{$licenseId}");
        return License::fromArray($data);
    }

    public function update(string $licenseId, array $input): License
    {
        $data = $this->client->patch("/sdk/licenses/{$licenseId}", $input);
        return License::fromArray($data);
    }

    public function delete(string $licenseId): void
    {
        $this->client->delete("/sdk/licenses/{$licenseId}");
    }
}

class MetricsResource
{
    public function __construct(private Management $client) {}

    public function list(string $licenseId, int $limit = 25, ?string $cursor = null): PaginatedResponse
    {
        $query = "?limit={$limit}";
        if ($cursor) {
            $query .= "&cursor={$cursor}";
        }
        $data = $this->client->get("/sdk/licenses/{$licenseId}/metrics{$query}");
        return new PaginatedResponse(
            data: array_map(fn($m) => Metric::fromArray($m), $data['data'] ?? []),
            nextCursor: $data['next_cursor'] ?? null,
            hasMore: $data['has_more'] ?? false,
        );
    }

    public function get(string $metricId): Metric
    {
        $data = $this->client->get("/sdk/metrics/{$metricId}");
        return Metric::fromArray($data);
    }

    public function create(string $licenseId, array $input): Metric
    {
        $data = $this->client->post("/sdk/licenses/{$licenseId}/metrics", $input);
        return Metric::fromArray($data);
    }

    public function update(string $metricId, array $input): Metric
    {
        $data = $this->client->patch("/sdk/metrics/{$metricId}", $input);
        return Metric::fromArray($data);
    }

    public function delete(string $metricId): void
    {
        $this->client->delete("/sdk/metrics/{$metricId}");
    }
}

class ExpensesResource
{
    public function __construct(private Management $client) {}

    public function list(
        int $limit = 25,
        ?string $cursor = null,
        ?string $appId = null,
        ?int $year = null,
        ?int $month = null,
    ): PaginatedResponse {
        $query = "?limit={$limit}";
        if ($cursor) $query .= "&cursor={$cursor}";
        if ($appId) $query .= "&app_id={$appId}";
        if ($year) $query .= "&year={$year}";
        if ($month) $query .= "&month={$month}";

        $data = $this->client->get("/sdk/expenses{$query}");
        return new PaginatedResponse(
            data: array_map(fn($e) => Expense::fromArray($e), $data['data'] ?? []),
            nextCursor: $data['next_cursor'] ?? null,
            hasMore: $data['has_more'] ?? false,
        );
    }

    public function get(string $expenseId): Expense
    {
        $data = $this->client->get("/sdk/expenses/{$expenseId}");
        return Expense::fromArray($data);
    }
}

File: sdks/swift/Sources/DoowTrack/Management.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public struct ManagementOptions {
    public var endpoint: String
    public var timeoutSeconds: Double
    public var retryCount: Int
    public var debug: Bool

    public init(
        endpoint: String = "https://api.doow.co",
        timeoutSeconds: Double = 30,
        retryCount: Int = 3,
        debug: Bool = false
    ) {
        self.endpoint = ProcessInfo.processInfo.environment["DOOW_TRACK_ENDPOINT"] ?? endpoint
        self.debug = ProcessInfo.processInfo.environment["DOOW_TRACK_DEBUG"] == "true" || debug
        self.timeoutSeconds = timeoutSeconds
        self.retryCount = retryCount
    }
}

public class Management {
    private let apiKey: String
    private let options: ManagementOptions
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder

    public lazy var apps = AppsResource(self)
    public lazy var contracts = ContractsResource(self)
    public lazy var licenses = LicensesResource(self)
    public lazy var metrics = MetricsResource(self)
    public lazy var expenses = ExpensesResource(self)

    public init(_ apiKey: String, options: ManagementOptions = ManagementOptions()) throws {
        let key = ProcessInfo.processInfo.environment["DOOW_TRACK_API_KEY"] ?? apiKey
        guard key.hasPrefix("dk_") else {
            throw DoowError("Invalid API key format. Must start with 'dk_'.")
        }
        self.apiKey = key
        self.options = options

        self.encoder = JSONEncoder()
        self.encoder.keyEncodingStrategy = .convertToSnakeCase
        self.encoder.dateEncodingStrategy = .iso8601

        self.decoder = JSONDecoder()
        self.decoder.keyDecodingStrategy = .convertFromSnakeCase
        self.decoder.dateDecodingStrategy = .iso8601
    }

    func request<T: Decodable>(_ method: String, _ path: String, body: Encodable? = nil) throws -> T {
        let url = URL(string: "\(options.endpoint.trimmingCharacters(in: CharacterSet(charactersIn: "/")))/sdk\(path)")!
        log("[doow/management] \(method) \(path)")

        for attempt in 0...options.retryCount {
            var request = URLRequest(url: url)
            request.httpMethod = method
            request.timeoutInterval = options.timeoutSeconds
            request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")

            if let body = body, method != "GET" && method != "DELETE" {
                request.httpBody = try encoder.encode(AnyEncodable(body))
            }

            let semaphore = DispatchSemaphore(value: 0)
            var responseData: Data?
            var responseError: Error?
            var httpResponse: HTTPURLResponse?

            let task = URLSession.shared.dataTask(with: request) { data, response, error in
                responseData = data
                responseError = error
                httpResponse = response as? HTTPURLResponse
                semaphore.signal()
            }
            task.resume()
            semaphore.wait()

            if let error = responseError {
                if attempt < options.retryCount {
                    Thread.sleep(forTimeInterval: pow(2, Double(attempt)))
                    continue
                }
                throw DoowError("Request failed: \(error.localizedDescription)")
            }

            if let status = httpResponse?.statusCode {
                if status >= 200 && status < 300 {
                    if let data = responseData, !data.isEmpty {
                        return try decoder.decode(T.self, from: data)
                    }
                    throw DoowError("Empty response")
                }

                if status >= 500 && attempt < options.retryCount {
                    Thread.sleep(forTimeInterval: pow(2, Double(attempt)))
                    continue
                }

                let errorBody = responseData.flatMap { String(data: $0, encoding: .utf8) } ?? ""
                throw DoowError("API error: \(errorBody)", statusCode: status)
            }
        }

        throw DoowError("Max retries exceeded")
    }

    func delete(_ path: String) throws {
        let _: EmptyResponse = try request("DELETE", path)
    }

    private func log(_ message: String) {
        if options.debug {
            fputs("\(message)\n", stderr)
        }
    }
}

struct EmptyResponse: Decodable {}

struct AnyEncodable: Encodable {
    let value: Encodable

    init(_ value: Encodable) {
        self.value = value
    }

    func encode(to encoder: Encoder) throws {
        try value.encode(to: encoder)
    }
}

public class AppsResource {
    private let mgmt: Management

    init(_ mgmt: Management) { self.mgmt = mgmt }

    public func list(cursor: String? = nil, limit: Int = 50) throws -> PaginatedResponse<App> {
        var path = "/apps?limit=\(limit)"
        if let cursor = cursor { path += "&cursor=\(cursor)" }
        return try mgmt.request("GET", path)
    }

    public func get(_ id: String) throws -> App {
        return try mgmt.request("GET", "/apps/\(id)")
    }

    public func create(_ input: CreateAppInput) throws -> App {
        return try mgmt.request("POST", "/apps", body: input)
    }

    public func delete(_ id: String) throws {
        try mgmt.delete("/apps/\(id)")
    }
}

public class ContractsResource {
    private let mgmt: Management

    init(_ mgmt: Management) { self.mgmt = mgmt }

    public func listByApp(_ appId: String, cursor: String? = nil, limit: Int = 50) throws -> PaginatedResponse<Contract> {
        var path = "/apps/\(appId)/contracts?limit=\(limit)"
        if let cursor = cursor { path += "&cursor=\(cursor)" }
        return try mgmt.request("GET", path)
    }

    public func get(_ id: String) throws -> Contract {
        return try mgmt.request("GET", "/contracts/\(id)")
    }

    public func create(_ appId: String, _ input: CreateContractInput) throws -> Contract {
        return try mgmt.request("POST", "/apps/\(appId)/contracts", body: input)
    }

    public func delete(_ id: String) throws {
        try mgmt.delete("/contracts/\(id)")
    }
}

public class LicensesResource {
    private let mgmt: Management

    init(_ mgmt: Management) { self.mgmt = mgmt }

    public func list(cursor: String? = nil, limit: Int = 50) throws -> PaginatedResponse<License> {
        var path = "/licenses?limit=\(limit)"
        if let cursor = cursor { path += "&cursor=\(cursor)" }
        return try mgmt.request("GET", path)
    }

    public func get(_ id: String) throws -> License {
        return try mgmt.request("GET", "/licenses/\(id)")
    }
}

public class MetricsResource {
    private let mgmt: Management

    init(_ mgmt: Management) { self.mgmt = mgmt }

    public func listByLicense(_ licenseId: String, cursor: String? = nil, limit: Int = 50) throws -> PaginatedResponse<Metric> {
        var path = "/licenses/\(licenseId)/metrics?limit=\(limit)"
        if let cursor = cursor { path += "&cursor=\(cursor)" }
        return try mgmt.request("GET", path)
    }

    public func get(_ id: String) throws -> Metric {
        return try mgmt.request("GET", "/metrics/\(id)")
    }

    public func create(_ licenseId: String, _ input: CreateMetricInput) throws -> Metric {
        return try mgmt.request("POST", "/licenses/\(licenseId)/metrics", body: input)
    }

    public func delete(_ id: String) throws {
        try mgmt.delete("/metrics/\(id)")
    }
}

public class ExpensesResource {
    private let mgmt: Management

    init(_ mgmt: Management) { self.mgmt = mgmt }

    public func listByApp(_ appId: String, cursor: String? = nil, limit: Int = 50) throws -> PaginatedResponse<Expense> {
        var path = "/apps/\(appId)/expenses?limit=\(limit)"
        if let cursor = cursor { path += "&cursor=\(cursor)" }
        return try mgmt.request("GET", path)
    }
}
